consul/server_manager/server_manager.go

```diff
@@ -0,0 +1,340 @@
+	return sm.serverConfigValue.Load().(serverConfig)
+}
+
+// New is the only way to safely create a new ServerManager struct.
+func New(logger *log.Logger, shutdownCh chan struct{}, clusterInfo ConsulClusterInfo) (sm *ServerManager) {
+	// NOTE(sean@): Can't pass *consul.Client due to an import cycle
+	sm = new(ServerManager)
+	sm.logger = logger
+	sm.clusterInfo = clusterInfo
+	sm.shutdownCh = shutdownCh
+
+	sc := serverConfig{}
+	sc.servers = make([]*server_details.ServerDetails, 0)
+	sm.saveServerConfig(sc)
+	return sm
+}
+
+// NotifyFailedServer marks the passed in server as "failed" by rotating it
+// to the end of the server list.
+func (sm *ServerManager) NotifyFailedServer(server *server_details.ServerDetails) {
+	serverCfg := sm.getServerConfig()
+
+	// If the server being failed is not the first server on the list,
+	// this is a noop.  If, however, the server is failed and first on
+	// the list, acquire the lock, retest, and take the penalty of moving
+	// the server to the end of the list.
+
+	// Use atomic.CAS to emulate a TryLock().
+	if len(serverCfg.servers) > 0 && serverCfg.servers[0] == server &&
+		atomic.CompareAndSwapInt32(&sm.notifyFailedBarrier, 0, 1) {
+		defer atomic.StoreInt32(&sm.notifyFailedBarrier, 0)
+
+		// Grab a lock, retest, and take the hit of cycling the first
+		// server to the end.
+		sm.serverConfigLock.Lock()
+		defer sm.serverConfigLock.Unlock()
+		serverCfg = sm.getServerConfig()
+
+		if len(serverCfg.servers) > 0 && serverCfg.servers[0] == server {
+			serverCfg.servers = serverCfg.cycleServer()
+			sm.saveServerConfig(serverCfg)
+		}
+	}
+}
+
+// NumServers takes out an internal "read lock" and returns the number of
+// servers.  numServers includes both healthy and unhealthy servers.
+func (sm *ServerManager) NumServers() (numServers int) {
+	serverCfg := sm.getServerConfig()
+	numServers = len(serverCfg.servers)
+	return numServers
+}
+
+// RebalanceServers takes out an internal write lock and shuffles the list of
+// servers on this agent.  This allows for a redistribution of work across
+// consul servers and provides a guarantee that the order of the server list
+// isn't related to the age at which the node was added to the cluster.
+// Elsewhere we rely on the position in the server list as a hint regarding
+// the stability of a server relative to its position in the server list.
+// Servers at or near the front of the list are more stable than servers near
+// the end of the list.
+func (sm *ServerManager) RebalanceServers() {
+	sm.serverConfigLock.Lock()
+	defer sm.serverConfigLock.Unlock()
+	serverCfg := sm.getServerConfig()
+
+	newServers := make([]*server_details.ServerDetails, len(serverCfg.servers))
+	copy(newServers, serverCfg.servers)
+
+	// Shuffle the server list on server join.  Servers are selected from
+	// the head of the list and are moved to the end of the list on
+	// failure.
+	for i := len(serverCfg.servers) - 1; i > 0; i-- {
```

Comment:
An `int`-typed `len()` always horrifies me as a former C++ programmer, but it makes this kind of thing more robust.